Added feature: Climate React Auto Setup plus minor Bugfix

seidnerj opened the pull request:
See README changes for more details.

seidnerj committed:
Added option to disable AUTO mode

This defaults to "false" and can be controlled via the plugin's configuratiuon.

seidnerj committed:
Added a Climate React Auto Setup feature

When enabled, every time an AC's state is set or changed, an appropriate Climate React configuration will be set up for maintaining the desired temperature.

See README for more details.

seidnerj committed:
Climate React (AKA "Smart Mode") related bug fix

If climate react has never been set up for a given device, Sensibo's API returns "null" for the "smartMode" property of the device which leads to an invalid HomeKit value set by the backend.

seidnerj committed:
Minor styling adjustments

seidnerj committed:
Updated README

seidnerj committed:
reverted inadvertent change

seidnerj commented:
@benwebbbenwebb ? @nitaybz ? Any takers?

benwebbbenwebb commented:
I’ll take a look over the weekend @seidnerj 

I do however suspect that there may be some conflicts with the v2.4 alpha branch which we’ll need to work through (it has code to disable AUTO)

seidnerj commented:
Great, much appreciated!

seidnerj committed:
Additional console output for debug mode

As well as maintain consistency in logging format between GET and SET methods.

seidnerj committed:
Minor styling adjustments

seidnerj commented:
There's a few fixes I'm adding to this pull request so no need to review it for now. I'll update this thread when everything's ready.

Thanks in advance!

seidnerj committed:
Added feature to enable Climate React Switch in Accessory

Also added some additional logging logic for debug mode as well as appropriate comments.

seidnerj committed:
Update config schema for relevant changes

benwebbbenwebb commented:
Hi @seidnerj 

I know you're still working on this but I took a look and noticed that there is likely going to be merge conflicts when I push my [development branch](https://github.com/nitaybz/homebridge-sensibo-ac/blob/development/README.md) (my fault for not merging it towards main/master much, much sooner... got lazy as I was the only one active in the repo) that some users are alpha testing.

There is also feature cross-over with your AUTO mode suppression, I went whole hog and setup an array so users could exclude any and all modes (see readme example in the link above). I was thinking of deprecating the use of _disableX-mode_ in the future.

That all said, I haven't done anything similar to your ClimateReact changes. So I'm wondering if I should do a sort of cherry-pick (it may need to be manual...) of those in to (my) development branch, and then probably push to main/master (without a release just yet), to get us in sync?

Keen for your thoughts on the best way to proceed!

seidnerj committed:
Additional work on "Climate React Switch" in Accessory

seidnerj committed:
small bug fix

seidnerj committed:
Minor styling adjustments

seidnerj commented:
Hey @benwebbbenwebb!
Sure - depending on the amount of changes in your branch, it might be easier for me (or you?) to merge those changes onto my fork and then push it to main/master.

On a related note, I only have one small issue that I'm dealing with but I can't seem to track it. The issue is that when the ClimateReact Switch changes state, this has no effect, the code reaches "StateManager.set.ClimateReactEnabledSwitch" but StateHandler.set never follows. 

Would love for you to take a look and give me a clue as to what is wrong. I've compared it to the Light Switch accessory which kinda has identical functionality (and does work), and I didn't notice any difference.

Thanks in advance!

benwebbbenwebb commented:
Hi @seidnerj 

> Sure - depending on the amount of changes in your branch, it might be easier for me (or you?) to merge those changes onto my fork and then push it to main/master.

[My branch compared to master](https://github.com/nitaybz/homebridge-sensibo-ac/compare/master..development), I made some linting changes which doesn't help with the comparison unfortunately, but there is quite a bit there.
[Your fork (master) compared to this repos master](https://github.com/nitaybz/homebridge-sensibo-ac/compare/master..seidnerj:homebridge-sensibo-ac:master)

I could do a merge/PR of my branch to master and then we check where that leaves this PR?

As for your question:
> On a related note, I only have one small issue that I'm dealing with but I can't seem to track it. The issue is that when the ClimateReact Switch changes state, this has no effect, the code reaches "StateManager.set.ClimateReactEnabledSwitch" but StateHandler.set never follows.

I didn't pull your full changes to check, but I did note that you have the exact same service name for the ClimateReactService in both the AirConditioner and the ClimateReactSwitch files, wondering if there is a clash. You might need to clear out all cached Accessories manually from Homebridge as a starting point to be sure.

I also noted that both reference the same ClimateReactEnabledSwitch get/set functions in StateManager.js... probably should be fine (but I've also not used the ClimateReact, LightSwitch or SyncButton services to know if there is a common issue when the same setter is referenced from multiple Accessories).

benwebbbenwebb commented:
Oh, and I think maybe you also missed adding changes to updateHomeKit within AirConditioner.js to push any change back to the Home app UI (similar to what is here: https://github.com/seidnerj/homebridge-sensibo-ac/blob/2a35793d6718be02c571e1083cbb6d57e0132f78/homekit/ClimateReactSwitch.js#L67)

seidnerj commented:
Hey @benwebbbenwebb,
Yes - if you could do a merge/PR of your branch to master that will be a good step towards merging this PR as well. Thanks!

Re you comments regarding having the same exact service name in two separate file / using same StateManager functions. I I followed the example set by the "LightSwitch" accessory implementation, which also uses the same name/functions. I haven't seen any issues that seem to be related to this, so I believe it is okay. 

Re the "updateHomeKit" function, really the only state that's reflected in the Home app is whether Climate React is enabled or not, which the current code already takes care off. I can't think of what else should be there.


> Hi @seidnerj
> 
> > Sure - depending on the amount of changes in your branch, it might be easier for me (or you?) to merge those changes onto my fork and then push it to main/master.
> 
> [My branch compared to master](https://github.com/nitaybz/homebridge-sensibo-ac/compare/master..development), I made some linting changes which doesn't help with the comparison unfortunately, but there is quite a bit there. [Your fork (master) compared to this repos master](https://github.com/nitaybz/homebridge-sensibo-ac/compare/master..seidnerj:homebridge-sensibo-ac:master)
> 
> I could do a merge/PR of my branch to master and then we check where that leaves this PR?
> 
> As for your question:
> 
> > On a related note, I only have one small issue that I'm dealing with but I can't seem to track it. The issue is that when the ClimateReact Switch changes state, this has no effect, the code reaches "StateManager.set.ClimateReactEnabledSwitch" but StateHandler.set never follows.
> 
> I didn't pull your full changes to check, but I did note that you have the exact same service name for the ClimateReactService in both the AirConditioner and the ClimateReactSwitch files, wondering if there is a clash. You might need to clear out all cached Accessories manually from Homebridge as a starting point to be sure.
> 
> I also noted that both reference the same ClimateReactEnabledSwitch get/set functions in StateManager.js... probably should be fine (but I've also not used the ClimateReact, LightSwitch or SyncButton services to know if there is a common issue when the same setter is referenced from multiple Accessories).

benwebbbenwebb commented:
Hi @seidnerj 

I've done the merge to master #112 and (as expected) there is unfortunately some merge conflicts :(

Let me know if I can be of any help resolving them

seidnerj commented:
No worries, I will take a look and let you know if I need any help 🙏

benwebbbenwebb commented:
> Re you comments regarding having the same exact service name in two separate file / using same StateManager functions. I I followed the example set by the "LightSwitch" accessory implementation, which also uses the same name/functions. I haven't seen any issues that seem to be related to this, so I believe it is okay.

I think you're probably correct on that...

> Re the "updateHomeKit" function, really the only state that's reflected in the Home app is whether Climate React is enabled or not, which the current code already takes care off. I can't think of what else should be there.

I still recommend you try adding the updateValue... at the moment it sounds like the `.on('set'`... isn't being fired (is that true from what you can see via debug logs?) and if so that's probably because the value isn't changing anywhere for StateHander to see it?

I'm also conscious that you are updating values one level deeper than all the other calls... e.g. device.state._smartMode_.**_enabled_** rather than device.state.**_pureBoost_**. Not sure if that is impactful, the state handling was all written by nitaybz long before I started tinkering, so maybe that's another place to check (and eliminate)?

seidnerj committed:
Merge (manual) with upstream/master (work in progress)

seidnerj commented:
I've done the merge with master manually but haven't tested it yet. I'm sure there's going to be issues. I will also remove the disable auto feature as this is now duplicate functionality. Will probably take me a few days to get this sorted. I'll update the thread when I have news. Thx.

seidnerj committed:
Manual merge continued (work in progress)

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -250,105 +325,111 @@ module.exports = (device, platform) => {
 
 			TargetHumidifierDehumidifierState: (callback) => {
-				log.easyDebug(device.name, '(GET) - Target Dehumidifier State: DEHUMIDIFIER')
+				log.easyDebug(device.name, '(GET) - HeaterCooler State: ?')
```
Log line needs to be checked

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				const status = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name, '(SET) - AC Swing:', swing)
```
Think this should be state as I don't think swing is defined
```suggestion
				log.easyDebug(device.name, '(SET) - AC Swing:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name, '(SET) - AC Swing:', swing)
+				device.state.swing = state
```
```suggestion
				device.state.verticalSwing = state
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			},
 
 			ACRotationSpeed: (speed, callback) => {
+				log.easyDebug(device.name + ' -> Setting AC Rotation Speed:', speed + '%')
```
```suggestion
				log.easyDebug(device.name, '(SET) - AC Rotation Speed:', speed + '%')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 				const mode = characteristicToMode(lastMode)
-
-				log.easyDebug(device.name + ' -> Setting AC Rotation Speed:', speed + '%')
 				log.easyDebug(device.name + ' -> Setting Mode to', mode)
```
```suggestion
				log.easyDebug(device.name, '(SET) - HeaterCooler State:', mode)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-
-				if (status) {
+				state = !!state
+				log.easyDebug(device.name + ' -> Setting Fan state Active:', state)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Fan state Active:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				state = !!state
+				log.easyDebug(device.name + ' -> Setting Fan state Active:', state)
+				if (state) {
+					device.state.active = true
```
Duplicate (see 3 lines below), remove
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				log.easyDebug(device.name + ' -> Setting Fan state Active:', state)
+				if (state) {
+					device.state.active = true
 					log.easyDebug(device.name + ' -> Setting Mode to: FAN')
```
```suggestion
					log.easyDebug(device.name, '(SET) - Mode to: FAN')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			FanSwing: (state, callback) => {
-				const status = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Fan Swing:', state)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Fan Swing:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				const status = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Fan Swing:', state)
+				device.state.swing = state
```
```suggestion
				device.state.verticalSwing = state
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Fan Swing:', state)
+				device.state.swing = state
 
```
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			},
 
 			FanRotationSpeed: (speed, callback) => {
 				log.easyDebug(device.name + ' -> Setting Fan Rotation Speed:', speed + '%')
```
```suggestion
				log.easyDebug(device.name, '(SET) - Fan Rotation Speed:', speed + '%')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				device.state.mode = 'FAN'
+
 				device.state.active = true
+				log.easyDebug(device.name + ' -> Setting Mode to: FAN')
```
```suggestion
				log.easyDebug(device.name, '(SET) - Mode to: FAN')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 
-				log.easyDebug(device.name + ' -> Setting Fan Swing:', status)
+				device.state.active = true
 				log.easyDebug(device.name + ' -> Setting Mode to: FAN')
```
```suggestion
				log.easyDebug(device.name, '(SET) - Mode to: FAN')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-
-				if (status) {
+				state = !!state
+				log.easyDebug(device.name + ' -> Setting Dry state Active:', state)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Dry state Active:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			DrySwing: (state, callback) => {
-				const status = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Dry Swing:', state)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Dry Swing:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				const status = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				state = state === Characteristic.SwingMode.SWING_ENABLED ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Dry Swing:', state)
+				device.state.swing = state
```
```suggestion
				device.state.verticalSwing = state
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			},
 
 			DryRotationSpeed: (speed, callback) => {
 				log.easyDebug(device.name + ' -> Setting Dry Rotation Speed:', speed + '%')
```
```suggestion
				log.easyDebug(device.name, '(SET) - Dry Rotation Speed:', speed + '%')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				log.easyDebug(device.name + ' -> Setting Horizontal Swing:', status)
-				device.state.horizontalSwing = status
+				state = state ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name + ' -> Setting Horizontal Swing Swing:', state)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Horizontal Swing Swing:', state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 				callback()
 			},
 
 			// AC LIGHT
```
This is common for AC and AirPurifer
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-
-				log.easyDebug(device.name + ' -> Setting AC Light to', lightState ? 'ON' : 'OFF')
-				device.state.light = lightState
+				log.easyDebug(device.name + ' -> Setting AC Light to', state ? 'ON' : 'OFF')
```
```suggestion
				log.easyDebug(device.name, '(SET) - Light to', state ? 'ON' : 'OFF')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 			SyncButton: (state, callback) => {
 				if (state) {
-					log.easyDebug(device.name + ' -> Syncing AC State => Setting ' + (device.state.active ? 'OFF' : 'ON') + ' state without sending commands')
+					log.easyDebug(device.name + ' (SYNC) - AC Active State:', device.state.active)
```
```suggestion
					log.easyDebug(device.name, '(SYNC) - AC Active State:', device.state.active)
```

benwebbbenwebb commented on config.schema.json:
```diff
@@ -81,7 +81,7 @@
 			"disableDry": {
 				"title": "Disable Dry Accessory",
-				"description": "Disable DRY mode control - remove dehumidifier accessory",
+				"description": "Disable DRY mode control - remove extra dehumidifier accessory",
```
Not clear on why it would be 'extra'?

benwebbbenwebb commented on config.schema.json:
```diff
@@ -123,7 +123,7 @@
 			"enableClimateReactSwitch": {
 				"title": "Enable Climate React Switch",
-				"description": "Adds a switch to enable/disable Climate React",
+				"description": "Adding a switch to quickly enable/disable Climate React",
```
Not sure about this sentence, other options are consistent with "Adds"

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -26,21 +26,26 @@ class AirConditioner {
 		this.disableDry = platform.disableDry
 		this.disableFan = platform.disableFan
-		this.disableHumidity = platform.disableHumidity
+		this.disableDry = platform.disableDry
```
See 2 rows above
```suggestion
```

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -26,21 +26,26 @@ class AirConditioner {
+		this.modesToExclude = platform.modesToExclude
 		this.temperatureUnit = deviceInfo.temperatureUnit
+		this.climateReactSwitchInAccessory = platform.climateReactSwitchInAccessory
+		this.disableAuto = platform.disableAuto
```
WIP - remove

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -26,21 +26,26 @@ class AirConditioner {
-		this.modesToExclude = platform.modesToExclude
 		this.filterService = deviceInfo.filterService
-		this.capabilities = unified.capabilities(device, platform)
+		this.capabilities = unified.capabilities(device, platform, this.disableAuto)
```
WIP - remove

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -26,21 +26,26 @@ class AirConditioner {
-		this.capabilities = unified.capabilities(device, platform)
+		this.capabilities = unified.capabilities(device, platform, this.disableAuto)
 
+		this.state = this.cachedState.devices[this.id] = unified.acState(device)
```
Duplicate, see 2 rows below
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -42,6 +42,66 @@ function sanitize(service, characteristic, value) {
+	device.state.smartMode.highTemperatureWebhook = null	
+	device.state.smartMode.lowTemperatureWebhook = null	
+
+	if (device.state.mode === 'COOL') {
```
Seems like there is a lot of duplication within this if/else... could you set once and then update/override just the relevant lines, e.g.
```
if (COOL) {
device.state.smartMode.highTemperatureThreshold = device.state.targetTemperature
device.state.smartMode.highTemperatureState.on
...etc
} else if (HEAT) {
device.state.smartMode.highTemperatureThreshold = device.state.targetTemperature + (device.usesFahrenheit ? 1.8 : 1)
device.state.smartMode.lowTemperatureState.on
...etc
}
```

seidnerj replied:
I'll get to this after the merge is done and working

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -42,6 +42,66 @@ function sanitize(service, characteristic, value) {
+			temperatureUnit: device.temperatureUnit,
+			mode: device.state.mode,
+			fanSpeed: device.state.fanSpeed,
+			swing: device.state.swing,
```
I moved to using verticalSwing
```suggestion
			swing: device.state.verticalSwing,
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -42,6 +42,66 @@ function sanitize(service, characteristic, value) {
+			temperatureUnit: device.temperatureUnit,				
+			mode: device.state.mode,
+			fanSpeed: device.state.fanSpeed,
+			swing: device.state.swing,
```
I moved to using verticalSwing
```suggestion
			swing: device.state.verticalSwing,
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -42,6 +42,66 @@ function sanitize(service, characteristic, value) {
+			temperatureUnit: device.temperatureUnit,
+			mode: device.state.mode,
+			fanSpeed: device.state.fanSpeed,
+			swing: device.state.swing,
```
I moved to using verticalSwing
```suggestion
			swing: device.state.verticalSwing,
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -42,6 +42,66 @@ function sanitize(service, characteristic, value) {
+			temperatureUnit: device.temperatureUnit,
+			mode: device.state.mode,
+			fanSpeed: device.state.fanSpeed,
+			swing: device.state.swing,
```
I moved to using verticalSwing
```suggestion
			swing: device.state.verticalSwing,
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -250,105 +325,111 @@ module.exports = (device, platform) => {
 				callback(null, horizontalSwing === 'SWING_ENABLED')
 			},
 
 			// AC LIGHT
```
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -250,105 +325,111 @@ module.exports = (device, platform) => {
 			// AC LIGHT
 			LightSwitch: (callback) => {
 				const light = device.state.light
+				log.easyDebug(device.name, '(GET) - AC Light:', light ? 'ON' : 'OFF')
```
LightSwitch is common...
```suggestion
				log.easyDebug(device.name, '(GET) - Light:', light ? 'ON' : 'OFF')
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				device.state.targetTemperature = targetTemp								
+				log.easyDebug(device.name, '(SET) - target HeaterCooler State:', mode)
 				device.state.mode = mode
-				device.state.active = true
```
Was this intentionally removed?

seidnerj replied:
yes, but I'll revert it.

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				device.state.targetTemperature = targetTemp
+				log.easyDebug(device.name, '(SET) - HeaterCooler State:', mode)
 				device.state.mode = mode
-				device.state.active = true
```
Was this intentionally removed?

seidnerj replied:
yes, but I'll revert it.

benwebbbenwebb commented on homekit/SyncButton.js:
```diff
@@ -53,7 +53,8 @@ class SyncButton {
 	}
 
 	addSyncButtonService() {
-		this.log.easyDebug(`${this.name} - Adding SyncButtonService`)
```
I've been trying to be consistent with this notation...

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -76,13 +76,22 @@ module.exports = {
 	},
 
-	capabilities: (device, platform) => {
+	capabilities: (device, platform, disableAuto=false) => {
```
WIP - remove disableAuto

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -76,13 +76,22 @@ module.exports = {
-	capabilities: (device, platform) => {
+	capabilities: (device, platform, disableAuto=false) => {
+
+		log = platform.log
```
Not required

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -76,13 +76,22 @@ module.exports = {
 			const mode = key.toUpperCase()
 
+			// if "disableAuto" is set, then skip adding AUTO mode.
+			if (disableAuto && mode === 'AUTO') {
```
WIP - remove disableAuto

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -196,6 +205,14 @@ module.exports = {
 		const modeCapabilities = device.remoteCapabilities.modes[device.acState.mode]
 
+		if (modeCapabilities.swing && modeCapabilities.swing.includes('rangeFull')) {
+			state.swing = device.acState.swing === 'rangeFull' ? 'SWING_ENABLED' : 'SWING_DISABLED'
```
```suggestion
			state.verticalSwing = device.acState.swing === 'rangeFull' ? 'SWING_ENABLED' : 'SWING_DISABLED'
```

benwebbbenwebb commented on index.js:
Additions / edits to this file need to be checked, lots of duplication of attributes:
- name
- disableFan
- disableDry
- enableHistoryStorage
- locationsToInclude
- disableHorizontalSwing
- disableVerticalSwing
- disableLightSwitch
- allowRepeatCommands

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -71,37 +76,39 @@ module.exports = (device, platform) => {
 			if (prop === 'smartMode') {
 				try {
-					log.easyDebug(`smartMode - updating ${device.name} to ${value}`)
+					log(`Setting Climate React state for ${device.name} to ${value}`)
```
```suggestion
					log.easyDedub(`Setting Climate React state for ${device.name} to ${value}`)
```

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -71,37 +76,39 @@ module.exports = (device, platform) => {
 			if (prop === 'pureBoost') {
 				try {
-					log.easyDebug(`pureBoost - updating ${device.name} to ${value}`)
+					log(`Setting Pure Boost state for ${device.name} to ${value}`)
```
```suggestion
					log.easyDebug(`Setting Pure Boost state for ${device.name} to ${value}`)
```

benwebbbenwebb commented on sensibo/api.js:
the newly added apiRequest and getToken functions are duplicates and should be removed

fixResponse can be kept and call it from the existing async apiRequest function

benwebbbenwebb reviewed (changes requested):
Hello @seidnerj 

I know you're still working on it but I've done a first pass and added some suggestions and comments.

Please take a look when you have a chance.

Thanks!

seidnerj committed:
Additional work on merge with upstream/master

seidnerj commented:
> Hello @seidnerj
> 
> I know you're still working on it but I've done a first pass and added some suggestions and comments.
> 
> Please take a look when you have a chance.
> 
> Thanks!

I've fixed all pertinent the issues you raised in this latest commit.

benwebbbenwebb commented:
> I've fixed all pertinent the issues you raised in this latest commit.

Thanks! Sorry for the hassle. 🙇 

Before I do another (maybe final?) review, please run a lint check. You may have already done this, if so, let me know.

`npm run lint` (or `npm run lint:fix` if you're happy for eslint to 'fix' it on the fly).

Also, did you figure out your issue mentioned above and successfully test your changes?

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -292,6 +302,43 @@ module.exports = {
+			lowTemperatureState:
+			{
+				on: smartModeState.lowTemperatureState.on,
+				swing: smartModeState.lowTemperatureState.swing === 'SWING_ENABLED' ? 'rangeFull' : 'stopped',
```
Where does .swing come from in smartModeState?

The Sensibo API can return a number of values for `swing`, including (but not limited to): rangeFull, horizontal, both, stopped... (plus several other "ranged" and "fixed" values like fixedTop, rangeMiddle etc). `horizontalSwing` is similar, but without 'horizontal' (or vertical) and with center etc instead of top...

Internally I started using state.verticalSwing and state.horizontalSwing to track them separately to account for the various options, and then convert back to how Sensibo 'expect' them just-in-time... Will something similar to line 274 above be required here? (Line 179 has the transformation the other way)

seidnerj replied:
unified.sensiboFormattedClimageReactState() just formats "state.smartMode" to Sensibo's format, based on the values of state.smartMode, which are updated by StateManager.updateClimateReact() which is in turn called every time a relevant AC setting is changed in StateManager (e.g. ACActive, PureActive, TargetHeaterCoolerState, CoolingThresholdTemperature, HeatingThresholdTemperature, ACSwing, ACRotationSpeed, PureRotationSpeed ...)

I updated unified.sensiboFormattedClimageReactState() with the same logic that exists in unified.sensiboFormattedACState(). This is a temporary code should be refactored before release since it has a lot of duplication (both within the function and with the "sensiboFormattedACState" function).

There is a question however if/what should be the value of "state.horizontalSwing" if the device is capable of a "threeDimensionalSwing". This might necessitate changes in both ac & climate formatting functions.

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -196,6 +198,14 @@ module.exports = {
 
 		const modeCapabilities = device.remoteCapabilities.modes[device.acState.mode]
 
+		if (modeCapabilities.swing && modeCapabilities.swing.includes('rangeFull')) {
```
I think this line and the 6 below it can be removed (pretty sure they are left over from the merge conflicts)

seidnerj replied:
done.

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -77,6 +77,8 @@ module.exports = {
 
 	capabilities: (device, platform) => {
+
+		log = platform.log
```
```suggestion
```

Shouldn't be required as all logging in this function uses platform.log.easyDebug(... )

seidnerj replied:
This suggestion is outdated, I already removed this.

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -108,12 +173,12 @@ module.exports = (device, platform) => {
 				if (!active || mode === 'FAN' || mode === 'DRY') {
 					const lastMode = device.HeaterCoolerService.getCharacteristic(Characteristic.TargetHeaterCoolerState).value
 
+
```
```suggestion
```

benwebbbenwebb commented on sensibo/unified.js:
```diff
@@ -292,6 +302,43 @@ module.exports = {
-	}
+	},
 
+	sensiboFormattedClimageReactState: (device, state) => {
```
```suggestion
	sensiboFormattedClimateReactState: (device, state) => {
```

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -119,14 +126,25 @@ module.exports = (device, platform) => {
-
+		
+				const sensiboNewACState = unified.sensiboFormattedACState(device, state)
+				const sensiboNewClimateReactState = unified.sensiboFormattedClimageReactState(device, state)
```
```suggestion
				const sensiboNewClimateReactState = unified.sensiboFormattedClimateReactState(device, state)
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 					device.state.active = false
 				}
 
+				updateClimateReact()
```
Wondering if there is a better way so that `updateClimateReact()` only gets called when required (e.g. when enableClimateReactAutoSetup is true and a value relevant to ClimateReact has changed...)

Maybe somewhere in StateHandler before setDeviceState is called (somewhere around your updated sensiboFormattedState call).

seidnerj replied:
Definitely. Because this is such a painful merge, and it's yet to have been tested even, I'm trying to not add complexity for now. We can do this after the merge, once we see everything works. At the end of the day, this simply updates the internal state within the plugin, if the enableClimateReactAutoSetup is False, this does not get POSTed to Sensibo's API so I think there's very little impact here doing this.

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+
 				device.state.active = true
+				log.easyDebug(device.name, '(SET) - Mode to: FAN')
+				device.state.mode = 'FAN'
```
```suggestion
				device.state.mode = 'FAN'

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				log.easyDebug(device.name, '(SET) - Fan Rotation Speed:', speed + '%')
 				device.state.fanSpeed = speed
-				device.state.mode = 'FAN'
+
```
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				device.state.verticalSwing = state
 				device.state.active = true
+				log.easyDebug(device.name, '(SET) - Mode to: FAN')
+				device.state.mode = 'FAN'
```
```suggestion
				device.state.mode = 'FAN'

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 					device.state.active = false
 				}
-
 				callback()
```
```suggestion

				callback()
```

seidnerj commented:
Re the linting - will do. I haven't yet tested the post-merge version and checked that everything works. 

Once I've done that, I will resolve the issue of the climate react switch not firing.

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				device.state.verticalSwing = status
 				device.state.mode = 'DRY'
-				device.state.active = true
 				callback()
```
```suggestion

				callback()
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+
 				device.state.active = true
+				log.easyDebug(device.name + ' -> Setting Mode to: DRY')
+				device.state.mode = 'DRY'
```
```suggestion
				device.state.mode = 'DRY'

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+				log.easyDebug(device.name, '(SET) - Dry Rotation Speed:', speed + '%')
 				device.state.fanSpeed = speed
-				device.state.mode = 'DRY'
+
```
```suggestion
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				device.state.horizontalSwing = status
+				state = state ? 'SWING_ENABLED' : 'SWING_DISABLED'
+				log.easyDebug(device.name, '(SET) - Horizontal Swing Swing:', state)
+				device.state.horizontalSwing = state
```
```suggestion
				device.state.horizontalSwing = state

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
-				log.easyDebug(device.name + ' -> Setting AC Light to', lightState ? 'ON' : 'OFF')
-				device.state.light = lightState
+				log.easyDebug(device.name, '(SET) - Light to', state ? 'ON' : 'OFF')
+				device.state.light = state
```
```suggestion
				device.state.light = state

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
+			ClimateReactEnabledSwitch: (state, callback) => {
+				log.easyDebug(device.name, '(SET) - Climate React Enabled Switch:', state)
+				device.state.smartMode.enabled = state
 				callback()
```
```suggestion

				callback()
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +441,255 @@ module.exports = (device, platform) => {
 				const pureBoost = !!state
-
-				log.easyDebug(device.name + ' -> Setting Target AirPurifier State (PURE BOOST) to', pureBoost ? 'enabled' : 'disabled')
+				log.easyDebug(`${device.name} (SET) - Pure Target State (Boost): ${pureBoost ? 'AUTO' : 'MANUAL'}`)
```
```suggestion
				log.easyDebug(device.name, '(SET) - Pure Target State (Boost):', pureBoost ? 'AUTO' : 'MANUAL')
```

seidnerj committed:
Linting

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -481,6 +488,29 @@ class AirConditioner {
 	}
 
+	addClimateReactService() {
+		this.log.easyDebug(`Adding Climate React Switch Service in the ${this.roomName}`)
```
```suggestion
		this.log.easyDebug(`${this.roomName} - Adding Climate React Switch Service`)
```

benwebbbenwebb commented on homekit/AirConditioner.js:
```diff
@@ -481,6 +488,29 @@ class AirConditioner {
+
+		if (ClimateReactService) {
+			// remove service
+			this.log.easyDebug(`Removing Climate React Switch Service from the ${this.roomName}`)
```
```suggestion
			this.log.easyDebug(`${this.roomName} - Removing Climate React Switch Service`)
```

benwebbbenwebb commented on config-sample.json:
```diff
@@ -6,9 +6,9 @@
-	"description": "This is an example configuration for the sensibo-ac homebridge plugin",
-	"platforms": [
-		{
+  "description": "This is an example configuration for the sensibo-ac homebridge plugin",
```
```suggestion
	"description": "This is an example configuration for the sensibo-ac homebridge plugin",
```

benwebbbenwebb commented on config-sample.json:
```diff
@@ -6,9 +6,9 @@
-	"platforms": [
-		{
+  "description": "This is an example configuration for the sensibo-ac homebridge plugin",
+  "platforms": [
```
```suggestion
	"platforms": [
```

benwebbbenwebb commented on config-sample.json:
```diff
@@ -6,9 +6,9 @@
-		{
+  "description": "This is an example configuration for the sensibo-ac homebridge plugin",
+  "platforms": [
+    {
```
```suggestion
		{
```

benwebbbenwebb commented on config-sample.json:
```diff
@@ -31,9 +31,11 @@
+			"debug": false,
+			"climateReactSwitchInAccessory": false,
+			"enableClimateReactAutoSetup": false
+    }
```
```suggestion
		}
```

benwebbbenwebb commented on config-sample.json:
```diff
@@ -31,9 +31,11 @@
+			"climateReactSwitchInAccessory": false,
+			"enableClimateReactAutoSetup": false
+    }
+  ],
```
```suggestion
	],
```

benwebbbenwebb commented on README.md:
```diff
@@ -109,25 +111,27 @@ See below the table for additional details on these settings.
+| `locationsToInclude`       |  Device location IDs or names to include when discovering Sensibo devices (leave empty for all locations)  |          |     -    |  String[]  |
+| `modesToExclude`           |  Modes to exclude from Home app when setting up Sensibo devices (leave empty to keep all available modes). Valid values: AUTO, COOL, DRY, FAN, HEAT  |          |     -    |  String[]  |
+| `debug`                    |  When set to `true`, the plugin will produce extra logs for debugging purposes  |          |  `false` |  Boolean  |
+| `climateReactSwitchInAccessory`         |  When set to `true`, it will remove the extra AC Climate React switch if it exists and will show \"Climate React Button\" attached as a service to the Same AC Accessory (works only when `enableClimateReactSwitch` is set to true)   |          |  `false` |  Boolean  |
```
```suggestion
| `climateReactSwitchInAccessory`         |  When set to `true`, adds a **Climate React Button** (like `enableClimateReactSwitch` above) but within the AC Accessory. It will also remove the standalone AC Climate React switch (if one exists). Works only when `enableClimateReactSwitch` is also set to true  |          |  `false` |  Boolean  |
```

benwebbbenwebb commented on README.md:
```diff
@@ -109,25 +111,27 @@ See below the table for additional details on these settings.
+| `modesToExclude`           |  Modes to exclude from Home app when setting up Sensibo devices (leave empty to keep all available modes). Valid values: AUTO, COOL, DRY, FAN, HEAT  |          |     -    |  String[]  |
+| `debug`                    |  When set to `true`, the plugin will produce extra logs for debugging purposes  |          |  `false` |  Boolean  |
+| `climateReactSwitchInAccessory`         |  When set to `true`, it will remove the extra AC Climate React switch if it exists and will show \"Climate React Button\" attached as a service to the Same AC Accessory (works only when `enableClimateReactSwitch` is set to true)   |          |  `false` |  Boolean  |
+| `enableClimateReactAutoSetup` |  Enables setting up an appropriate Climate React configuration when AC state is set or changed.     |          |  `false` |  Boolean  |
```
```suggestion
| `enableClimateReactAutoSetup` |  When set to `true`, the plugin (whenever the AC state is set or changed) will update the Climate React configuration to match the new settings  |          |  `false` |  Boolean  |
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
+**To attach the "Climate React" button as a service to the AC accessory instead of a separate switch**, add 
+`"climateReactSwitchInAccessory": true` to your config.
+
+**To enable the a Climate React switch**, add `"enableClimateReactSwitch": true` to your config.
```
```suggestion
To enable the **Climate React switch**, add `"enableClimateReactSwitch": true` to your config.

To attach the **Climate React switch** as a service within the AC accessory, instead of a separate switch, also add `"climateReactSwitchInAccessory": true` to your config.
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
 Note: This feature does not allow changing the actual logic (temperature, mode etc) of the "Climate React" mode, only enabling or disabling it. Therefore, it will not work if the "Climate React" was not set up in Sensibo app first.
 
-To enable the extra **Climate React** switch, add `"enableClimateReactSwitch": true` to your config.
+**To attach the "Climate React" button as a service to the AC accessory instead of a separate switch**, add 
```
```suggestion
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
 
-To enable the extra **Climate React** switch, add `"enableClimateReactSwitch": true` to your config.
+**To attach the "Climate React" button as a service to the AC accessory instead of a separate switch**, add 
+`"climateReactSwitchInAccessory": true` to your config.
```
```suggestion
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
-To enable the extra **Climate React** switch, add `"enableClimateReactSwitch": true` to your config.
+**To attach the "Climate React" button as a service to the AC accessory instead of a separate switch**, add 
+`"climateReactSwitchInAccessory": true` to your config.
+
```
```suggestion
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
+**To enable the a Climate React switch**, add `"enableClimateReactSwitch": true` to your config.
+
+### Climate React Auto Setup
+When enabled, every time an AC's state is set or changed,  appropriate Climate React configuration will be set up for maintaining the desired temperature.
```
```suggestion
When enabled, every time an AC's state is set or changed, the Climate React configuration will be updated to maintaining the desired temperature.
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
+### Climate React Auto Setup
+When enabled, every time an AC's state is set or changed,  appropriate Climate React configuration will be set up for maintaining the desired temperature.
+
+For example, if setting an AC to Cool to 25C, Climate React will be set up such that when the temperature rises above 25C the AC starts to cool and when the temperature drops below 24 (the target temperature minus 1 degress centigrade, or the equivalent fahrenheit delta), the AC will be turned off.
```
```suggestion
For example, if setting an AC to Cool to 25C, Climate React will be set up such that when the temperature rises above 25C the AC starts to cool and when the temperature drops below 24 (the target temperature minus 1 degree C, or the equivalent F delta), the AC will be turned off.
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
+When setting an AC to Heat with a target temprature, Climate React will be set up similarly.
+
+**To enable Climate React Auto Setup**, add 
+`"enableClimateReactAutoSetup": true` to your config.
```
```suggestion
```

benwebbbenwebb commented on README.md:
```diff
@@ -219,15 +223,28 @@ Note: Geofencing must be enabled in Sensibo app for it to work
+
+When setting an AC to Heat with a target temprature, Climate React will be set up similarly.
+
+**To enable Climate React Auto Setup**, add 
```
```suggestion
**To enable Climate React Auto Setup**, add `"enableClimateReactAutoSetup": true` to your config.
```

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -71,33 +75,37 @@ module.exports = (device, platform) => {
 			if (prop === 'smartMode') {
 				try {
-					log.easyDebug(`smartMode - updating ${device.name} to ${value}`)
+					log.easyDebug(`Setting Climate React state for ${device.name} to ${value}`)
```
```suggestion
					log.easyDebug(`${device.name} - Setting Climate React state to ${value}`)
```

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -71,33 +75,37 @@ module.exports = (device, platform) => {
 			if (prop === 'pureBoost') {
 				try {
-					log.easyDebug(`pureBoost - updating ${device.name} to ${value}`)
+					log.easyDebug(`Setting Pure Boost state for ${device.name} to ${value}`)
```
```suggestion
					log.easyDebug(`${device.name} - Setting Pure Boost state to ${value}`)
```

benwebbbenwebb commented on sensibo/syncHomeKitCache.js:
```diff
@@ -218,9 +218,10 @@ module.exports = (platform) => {
 				case 'ClimateReact':
 					deviceExists = platform.devices.find(device => {
-						return device.id === accessory.context.deviceId && device.remoteCapabilities
+						device.id === accessory.context.deviceId && device.remoteCapabilities
```
I think this still needs to return for the .find to work?
```suggestion
						return device.id === accessory.context.deviceId && device.remoteCapabilities
```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -250,105 +325,120 @@ module.exports = (device, platform) => {
 				const airQuality = device.state.airQuality
 
-				log.easyDebug(device.name, '(GET) - Air Quality is:', airQuality)
+				log.easyDebug(device.name, '(GET) - Air Quality:', airQuality)
```
```suggestion
				log.easyDebug(device.name, '(GET) - Air Quality:', airQuality)

```

benwebbbenwebb commented on homekit/StateManager.js:
```diff
@@ -360,246 +450,263 @@ module.exports = (device, platform) => {
 					device.state.active = false
 				}
-
 				callback()
```
```suggestion

				callback()
```

benwebbbenwebb reviewed (commented):
Thanks for doing the linting @seidnerj

Few more suggestions from me here, once you've taken a look do you want me to merge to simplify your future changes or wait until you've completed the testing (and any bug squashing)?

seidnerj committed:
Continued work on merge with upstream/master

seidnerj commented:
I did another pass and committed the changes. 

Merging this PR to the current master branch should be relatively easy now. Merging your code with mine to create this PR was a much bigger issue. So, re your question, I think you can go ahead and merge this PR to the master branch and we can continue testing there then after completion we could do a release.

Once you've done the merge, I would appreciate any help testing this as this has become a much bigger deal than I expected (I wasn't expecting this mega merge) and I'm not sure when I'll be able to complete this.

seidnerj commented:
@benwebbbenwebb, any progress on merging this to master? thanks!

benwebbbenwebb commented:
Hi @seidnerj 

Apologies for the delay, misunderstood previous comments and was waiting for any additional changes on ClimateReact.

Merged now.

seidnerj commented:
Awesome, thanks! I'll get crackin'

seidnerj commented:
Okay, so - I tested the current master. The good news is that everything pretty much seems to work. The bad news is that the climate react enable/disable switch still doesn't work.

Here's a comparison of the behavior from the logs:

This is what happens when the AC is turned on:
```
[05/11/2023, 14:08:01] [homebridge-sensibo-ac] Living Room AC (SET) - AC Active State: true
[05/11/2023, 14:08:01] [homebridge-sensibo-ac] StateHandler SET active true for ...
```

This is what happens when climate react is turned on:
```
[05/11/2023, 14:07:32] [homebridge-sensibo-ac] Living Room AC (SET) - Climate React Enabled Switch: true
```

The line where it should say "StateHandler SET ...." is missing. It seems that it just doesn't get to StateHandler.set for some reason. I compared the existing code to the previous version and I couldn't find any change that would cause this behavior. I'm starting to wonder whether or not this worked before I got to work on this PR.

For the love of me, I can't figure out what's the problem there. Any help would be appreciated! 🙏

seidnerj commented:
@benwebbbenwebb any chance you could take a peek re the climate react switch issue?

benwebbbenwebb commented:
Hey @seidnerj 
Will do, might take a day or so but I’ll let you know (will also compare with v2.3.4)

benwebbbenwebb commented:
> @benwebbbenwebb any chance you could take a peek re the climate react switch issue?

Hi @seidnerj 

Can confirm that on v2.4.0-alpha.5 (what my main Homebridge is running currently) ClimateReact works in Home app and I can see the status update (e.g. it turns ClimateReact on/off) in the Sensibo app.

Will check on your implementation of the switch as a Service attached to the AC when I can.

seidnerj commented:
Great, many thanks!

benwebbbenwebb commented:
Hi @seidnerj

Please see 0536481, I'm not really sure _why_ (it was implemented before I took over), but looks like state change detection doesn't work 2 levels deep (`smartMode.enabled`). Keeping it at a single level (`smartModeEnabled`) is firing the `set` in StateHandler as expected...

Not sure what that means for the rest of your implementation, but it should allow you to progress at least.

Note, in my commit I didn't make changes to things like `prop === 'smartMode'` (line 79 of StateHandler), but when I tested it locally (by changing it to `smartModeEnabled`) the API call was triggered and Sensibo app updated.

Will leave it with you for now!

seidnerj commented:
I think I got why! thanks a lot, I'll take over - much appreciated 🙏

seidnerj commented:
Finally fixed this issue. I also documented in a comment what was the issue. I'l open a separate PR for you to review.
